### PR TITLE
Add cmake support

### PR DESCRIPTION
**Pull request by banshee**
I added cmake support for myself (I wanted to use CLion), and I'm happy to add it in to the project if it's interesting for other people.  There's some cleanup work to be done here before merge (although it's working), I just wanted to see if it's worth doing.

### CODE REVIEW

**Comment by albertschulz** [comment]
Would love to see CMake support for this repository. @aclements could this request be accepted?

**Comment by banshee** [comment]
Just FYI, it's been a long time since I touched anything to do with CMake, so I'm unlikely to be able to do contribute anything additional here.  I had totally forgotten about this until I saw the update from a week ago.

**Comment by banshee** [comment]
Looks like there's another PR that includes cmake - https://github.com/aclements/libelfin/pull/37

**Comment by Smertig** [comment]
I wish this PR (or #37) will be accepted. CMake is the most popular tool for managing C++ build process. 

@aclements, it would be great, if you take a look at this PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 3.6)
 
 project (elfin)
 
+set(CMAKE_CXX_STANDARD 17)
```
Comment (Smertig) [inline CMakeLists.txt:5]: I think it should be 11, not 17

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.6)
+
+project (elfin)
+
+set(CMAKE_CXX_STANDARD 17)
```
Comment (madebr) [inline CMakeLists.txt:5]:
```suggestion
set(CMAKE_CXX_STANDARD 11)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+add_subdirectory(elf elfdir)
+add_subdirectory(dwarf dwarfdir)
+add_subdirectory(examples examplesdir)
```
Comment (madebr) [inline CMakeLists.txt:9]: Adding an immediate subdirectory in the source tree as a different directory in the build directory is not necessary.
```suggestion
add_subdirectory(elf)
add_subdirectory(dwarf)
add_subdirectory(examples)
```

File: dwarf/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.6)
+
+project(dwarf)
+
+set(dwarfH ${CMAKE_CURRENT_LIST_DIR}/dwarf++.hh)
+set(dataH ${CMAKE_CURRENT_LIST_DIR}/data.hh)
```
Comment (madebr) [inline dwarf/CMakeLists.txt:6]:
```suggestion
set(dwarfH ${CMAKE_CURRENT_SOURCE_DIR}/dwarf++.hh)
set(dataH ${CMAKE_CURRENT_SOURCE_DIR}/data.hh)
```

File: dwarf/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+add_custom_command(
+  OUTPUT ${to_string_src}
+  COMMAND bash ${configured_to_string_script} ${dwarfH} ${dataH} ${to_string_src})
+add_custom_target(to_string_target DEPENDS ${to_string_src})
+
+add_library(dwarf dwarf.cc cursor.cc die.cc value.cc abbrev.cc
+  expr.cc rangelist.cc line.cc attrs.cc
+  die_str_map.cc elf.cc ${to_string_src})
+target_include_directories(dwarf PUBLIC ${CMAKE_CURRENT_LIST_DIR})
```
Comment (madebr) [inline dwarf/CMakeLists.txt:20]:
```suggestion
target_include_directories(dwarf PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
```

File: dwarf/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+set(dwarfH ${CMAKE_CURRENT_LIST_DIR}/dwarf++.hh)
+set(dataH ${CMAKE_CURRENT_LIST_DIR}/data.hh)
+set(to_string_src ${CMAKE_CURRENT_BINARY_DIR}/to_string.cc)
+set(configured_to_string_script ${CMAKE_CURRENT_BINARY_DIR}/create_to_string_code.sh)
+
+configure_file(create_to_string_code.sh ${configured_to_string_script})
+
+add_custom_command(
+  OUTPUT ${to_string_src}
+  COMMAND bash ${configured_to_string_script} ${dwarfH} ${dataH} ${to_string_src})
```
Comment (madebr) [inline dwarf/CMakeLists.txt:14]: bash is not really cross platform (it is less often available on Windows then python).
I would just use a `add_custom_command` with all commands from the bash script.

See https://github.com/conan-io/conan-center-index/pull/3283 for an example.

File: elf/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.6)
+
+project(elf)
+
+set(toStringInputHeader ${CMAKE_CURRENT_LIST_DIR}/data.hh)
```
Comment (madebr) [inline elf/CMakeLists.txt:5]:
```suggestion
set(toStringInputHeader ${CMAKE_CURRENT_SOURCE_DIR}/data.hh)
```

File: elf/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+set(to_string_src ${CMAKE_CURRENT_BINARY_DIR}/to_string.cc)
+set(configured_to_string_script ${CMAKE_CURRENT_BINARY_DIR}/create_to_string_code.sh)
+
+configure_file(create_to_string_code.sh ${configured_to_string_script})
+
+add_custom_command(
+  OUTPUT ${to_string_src}
```
Comment (madebr) [inline elf/CMakeLists.txt:12]: Same remark as the dwarf library.
I would not use bash but add these commands directly here.
See https://github.com/conan-io/conan-center-index/pull/3283 again

File: elf/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  COMMAND bash ${configured_to_string_script} ${toStringInputHeader} ${to_string_src})
+add_custom_target(elf_to_string_target DEPENDS ${to_string_src})
+add_library(elf elf.cc mmap_loader.cc ${to_string_src})
+add_dependencies(elf elf_to_string_target)
```
Comment (madebr) [inline elf/CMakeLists.txt:16]: This dependency (or `add_custom_target`) should not be necessary.
As the generated source is added to the sources of the `add_library`, cmake knows already that the generated source is a dependency of the library and that its source should be generated first.

**Comment by madebr** [comment]
In addition to my remarks, cmake should also be tested on `.travis.yml`.

@banshee
Do you intend to update your pr?

**Comment by banshee** [comment]
No, it's been two years since I touched this and I haven't written anything in CMake for, hmm, can't actually remember.  I'm totally happy to do anything to help (that doesn't involve relearning CMake...), but my suggestion would be to take the file you ended up with and make it a new PR.  Abandoning this PR would be totally fine with me.

**Comment by banshee** [comment]
My next suggestion - if you want to put together a better PR that other people could also pull in, I'll happily close this one in favor of yours.  Things like "err, bash isn't exactly portable" are 100% true, and I think my original CMake files were pretty much the fastest thing I could get working.  Given that it's been here for two years, I suspect it's never going to make it into the main line, but it'd probably be good to only have one outstanding PR that people can grab.

**Comment by madebr** [comment]
@banshee 
Thanks for the reply!
I would love to propose a cmake pr myself,
but I will not invest any time in this until @aclements (or someone who has commit access) comments on this.

**Comment by banshee** [comment]
Closing this so it won't show up in my PR list any more.

**Pull request closed without merging**